zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -563,46 +563,49 @@ mod tests {
             )"#,
             // Internal table created by SQLite when we started using `AUTOINCREMENT`.
             "CREATE TABLE sqlite_sequence(name,seq)",
-            "CREATE TABLE transactions (
+            r#"CREATE TABLE "transactions" (
                 id_tx INTEGER PRIMARY KEY,
                 txid BLOB NOT NULL UNIQUE,
                 created TEXT,
                 block INTEGER,
+                mined_height INTEGER,
                 tx_index INTEGER,
                 expiry_height INTEGER,
                 raw BLOB,
                 fee INTEGER,
-                FOREIGN KEY (block) REFERENCES blocks(height)
-            )",
-            "CREATE TABLE transparent_received_output_spends (
+                FOREIGN KEY (block) REFERENCES blocks(height),
+                CONSTRAINT height_consistency CHECK (block IS NULL OR mined_height = block)
+            )"#,
+            r#"CREATE TABLE "transparent_received_output_spends" (
                 transparent_received_output_id INTEGER NOT NULL,
                 transaction_id INTEGER NOT NULL,
                 FOREIGN KEY (transparent_received_output_id)
-                    REFERENCES utxos(id)
+                    REFERENCES transparent_received_outputs(id)
                     ON DELETE CASCADE,
                 FOREIGN KEY (transaction_id)
                     -- We do not delete transactions, so this does not cascade
                     REFERENCES transactions(id_tx),
                 UNIQUE (transparent_received_output_id, transaction_id)
+            )"#,
+            "CREATE TABLE transparent_received_outputs (
+                id INTEGER PRIMARY KEY,
+                transaction_id INTEGER NOT NULL,
+                output_index INTEGER NOT NULL,
+                account_id INTEGER NOT NULL,
+                address TEXT NOT NULL,
+                script BLOB NOT NULL,
+                value_zat INTEGER NOT NULL,
+                max_observed_unspent_height INTEGER,
```

Comment:
Is it clear what this means? I'm concerned about off-by-one errors.

From how it's used, it seems to mean that the TXO was either created in this block or unspent at the *start* of the block. Either that or we are incorrectly assuming that UTXOs are not spent in the block where they're created.

Reply:
The is the maximum block height at which we have observed that the TXO appears in the UTXO set.